framework/fp_vectors/Floats.h

```diff
@@ -150,32 +197,62 @@ struct Float16
     static constexpr Float16 quiet_NaN()        { return Float16(Holder{0x7e00}); }
     static constexpr Float16 signaling_NaN()    { return Float16(Holder{0x7d00}); }
 
+    constexpr inline bool     is_negative() const         { return sign != 0; }
+    constexpr inline bool     is_zero() const             { return (exponent == FLOAT16_DENORM_EXPONENT) && (mantissa == 0); }
+    constexpr inline bool     is_denormal() const         { return (exponent == FLOAT16_DENORM_EXPONENT) && (mantissa != 0); }
+
+    // NaNs
+    constexpr inline bool     is_general_nan() const      { return exponent == FLOAT16_NAN_EXPONENT; }
+    constexpr inline bool     is_inf() const              { return is_general_nan() && (mantissa == 0); }
+    constexpr inline bool     is_nan() const              { return is_general_nan() && (mantissa != 0); }
+    constexpr inline bool     is_snan() const             { return is_nan() && ((mantissa & FLOAT16_MANTISSA_QUIET_NAN_MASK) == 0); }
+    constexpr inline bool     is_qnan() const             { return is_nan() && ((mantissa & FLOAT16_MANTISSA_QUIET_NAN_MASK) != 0); }
+
+    constexpr inline uint16_t get_nan_payload() const     { return mantissa & (~FLOAT16_MANTISSA_QUIET_NAN_MASK); }
+
 private:
     struct Holder { uint16_t payload; };
     explicit constexpr Float16(Holder h) : as_hex(h.payload) {}
 #endif
 };
-static_assert(sizeof(struct Float16) == 2, "Float16 structure is not of the correct size");
+typedef struct Float16 Float16;
+
+// C interface
+static inline bool     Float16_is_negative(Float16 f)         { return f.sign != 0; }
+static inline bool     Float16_is_zero(Float16 f)             { return (f.exponent == FLOAT16_DENORM_EXPONENT) && (f.mantissa == 0); }
+static inline bool     Float16_is_denormal(Float16 f)         { return (f.exponent == FLOAT16_DENORM_EXPONENT) && (f.mantissa != 0); }
+
+static inline bool     Float16_is_general_nan(Float16 f)      { return f.exponent == FLOAT16_NAN_EXPONENT; }
```

Comment:
I'm not liking this name.  It covers infinity, which is a number.
Maybe something like:  is_inf_or_nan(f) or exponent_all_ones(f) or something.

Though this can be done as separate PR.  No need to block the PR for this.
If you feel like updating it - i can re-approve.